In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
from sklearn.model_selection import train_test_split
from pycaret.classification import *
import shap

In [3]:
raw_data_joined = pd.read_csv('./output/raw_data_joined.csv', index_col=0)

In [4]:
comorb_lst = [i for i in raw_data_joined.columns if "DISEASE" in i]
comorb_lst.extend(["HTN", "IMMUNOCOMPROMISED", "OTHER"])

demo_lst = [i for i in raw_data_joined.columns if "AGE_" in i]
demo_lst.append("GENDER")


vitalSigns_lst = raw_data_joined.iloc[:,193:-2].columns.tolist()

lab_lst = raw_data_joined.iloc[:,13:193].columns.tolist()

As warned by the authors ("Beware NOT to use the data when the target variable is present, as it is unknown the order of the event (maybe the target event happened before the results were obtained)") which means that the inpatient's lab results could have been obtained after the inpatient's entry in the ICU thus rendering that data pointless

We will focus on creating a model that works within the 0-2 window since it's more clinically relevant. Therefore, all the data from patients admitted between the 0-2 window should be dropped

In [5]:
# ICU per window
pd.crosstab(raw_data_joined.WINDOW, raw_data_joined.ICU)

ICU         0    1
WINDOW            
0-2       353   32
2-4       326   59
4-6       286   99
6-12      255  130
ABOVE_12  190  195

In [6]:
# Dropping the patients which were admitted to ICU in the first window
train_data = raw_data_joined.loc[~((raw_data_joined['WINDOW'] == '0-2') & (raw_data_joined['ICU'] == 1))]
train_data

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  AGE_PERCENTIL  GENDER  \
0                            0            1             60       0   
1                            0            1             60       0   
2                            0            1             60       0   
3                            0            1             60       0   
4                            0            1             60       0   
...                        ...          ...            ...     ...   
1920                       384            0             50       1   
1921                       384            0             50       1   
1922                       384            0             50       1   
1923                       384            0             50       1   
1924                       384            0             50       1   

      DISEASE GROUPING 1  DISEASE GROUPING 2  DISEASE GROUPING 3  \
0                    0.0                 0.0                 0.0   
1                    0.0                 0.0                 0.0   
2                    0.0                 0.0                 0.0   
3                    0.0                 0.0                 0.0   
4                    0.0                 0.0                 0.0   
...                  ...                 ...                 ...   
1920                 0.0                 0.0                 0.0   
1921                 0.0                 0.0                 0.0   
1922                 0.0                 0.0                 0.0   
1923                 0.0                 0.0                 0.0   
1924                 0.0                 0.0                 1.0   

      DISEASE GROUPING 4  DISEASE GROUPING 5  DISEASE GROUPING 6  ...  \
0                    0.0                 1.0                 1.0  ...   
1                    0.0                 1.0                 1.0  ...   
2                    0.0                 1.0                 1.0  ...   
3                    0.0                 1.0                 1.0  ...   
4                    0.0                 1.0                 1.0  ...   
...                  ...                 ...                 ...  ...   
1920                 0.0                 0.0                 0.0  ...   
1921                 0.0                 0.0                 0.0  ...   
1922                 0.0                 0.0                 0.0  ...   
1923                 0.0                 0.0                 0.0  ...   
1924                 0.0                 0.0                 0.0  ...   

      TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  \
0            -1.000000               -1.000000   
1            -1.000000               -1.000000   
2            -1.000000               -1.000000   
3            -1.000000               -1.000000   
4            -0.238095               -0.818182   
...                ...                     ...   
1920         -1.000000               -1.000000   
1921         -1.000000               -1.000000   
1922         -1.000000               -1.000000   
1923         -1.000000               -1.000000   
1924         -0.547619               -0.838384   

      BLOODPRESSURE_DIASTOLIC_DIFF_REL  BLOODPRESSURE_SISTOLIC_DIFF_REL  \
0                            -1.000000                        -1.000000   
1                            -1.000000                        -1.000000   
2                            -1.000000                        -1.000000   
3                            -1.000000                        -1.000000   
4                            -0.389967                         0.407558   
...                                ...                              ...   
1920                         -1.000000                        -1.000000   
1921                         -1.000000                        -1.000000   
1922                         -1.000000                        -1.000000   
1923                         -1.000000                        -1.000000   
1924                         -0.701863       

In [7]:
icu_above_2 = train_data.groupby('PATIENT_VISIT_IDENTIFIER')\
    .agg({'ICU': max})\
    .reset_index()\
    .rename(columns={'ICU': 'ICU_NEW'})
    
# Merge back to original df
training_data = train_data.merge(icu_above_2, on=['PATIENT_VISIT_IDENTIFIER'], how='left')

It is important to notice the addition of a new column at the end of the dataset that tells us if the patient went into the ICU after the window we are studying to obtain the groundtruth value

In [8]:
training_data.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  AGE_PERCENTIL  GENDER  \
0                         0            1             60       0   
1                         0            1             60       0   
2                         0            1             60       0   
3                         0            1             60       0   
4                         0            1             60       0   

   DISEASE GROUPING 1  DISEASE GROUPING 2  DISEASE GROUPING 3  \
0                 0.0                 0.0                 0.0   
1                 0.0                 0.0                 0.0   
2                 0.0                 0.0                 0.0   
3                 0.0                 0.0                 0.0   
4                 0.0                 0.0                 0.0   

   DISEASE GROUPING 4  DISEASE GROUPING 5  DISEASE GROUPING 6  ...  \
0                 0.0                 1.0                 1.0  ...   
1                 0.0                 1.0                 1.0  ...   
2                 0.0                 1.0                 1.0  ...   
3                 0.0                 1.0                 1.0  ...   
4                 0.0                 1.0                 1.0  ...   

   OXYGEN_SATURATION_DIFF  BLOODPRESSURE_DIASTOLIC_DIFF_REL  \
0               -1.000000                         -1.000000   
1               -1.000000                         -1.000000   
2               -1.000000                         -1.000000   
3               -1.000000                         -1.000000   
4               -0.818182                         -0.389967   

   BLOODPRESSURE_SISTOLIC_DIFF_REL  HEART_RATE_DIFF_REL  \
0                        -1.000000            -1.000000   
1                        -1.000000            -1.000000   
2                        -1.000000            -1.000000   
3                        -1.000000            -1.000000   
4                         0.407558            -0.230462   

   RESPIRATORY_RATE_DIFF_REL  TEMPERATURE_DIFF_REL  \
0                  -1.000000             -1.000000   
1                  -1.000000             -1.000000   
2                  -1.000000             -1.000000   
3                  -1.000000             -1.000000   
4                   0.096774             -0.242282   

   OXYGEN_SATURATION_DIFF_REL    WINDOW  ICU  ICU_NEW  
0                   -1.000000       0-2    0        1  
1                   -1.000000       2-4    0        1  
2                   -1.000000       4-6    0        1  
3                   -1.000000      6-12    0        1  
4                   -0.814433  ABOVE_12    1        1  

[5 rows x 232 columns]

In [9]:
# Valid cases per window
pd.crosstab(training_data.WINDOW, training_data.ICU)

ICU         0    1
WINDOW            
0-2       353    0
2-4       326   59
4-6       286   99
6-12      255  130
ABOVE_12  190  195

In [10]:
# keep only features from 0-2 hour window
mask_02 = training_data.WINDOW == '0-2'
training_data = training_data.loc[mask_02]

# valid cases for training
pd.crosstab(training_data.WINDOW, training_data.ICU_NEW)

ICU_NEW    0    1
WINDOW           
0-2      190  163

We are left with 353 rows total but is as expected taking into account the demand for clinically relevant results

In [12]:
training_data

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  AGE_PERCENTIL  GENDER  \
0                            0            1             60       0   
9                            2            0             10       0   
14                           3            0             40       1   
19                           4            0             10       0   
24                           5            0             10       0   
...                        ...          ...            ...     ...   
1868                       380            0             40       1   
1873                       381            1             90       0   
1878                       382            0             50       0   
1883                       383            0             40       1   
1888                       384            0             50       1   

      DISEASE GROUPING 1  DISEASE GROUPING 2  DISEASE GROUPING 3  \
0                    0.0                 0.0                 0.0   
9                    0.0                 0.0                 0.0   
14                   0.0                 0.0                 0.0   
19                   0.0                 0.0                 0.0   
24                   0.0                 0.0                 0.0   
...                  ...                 ...                 ...   
1868                 0.0                 0.0                 0.0   
1873                 0.0                 0.0                 0.0   
1878                 0.0                 0.0                 0.0   
1883                 0.0                 0.0                 0.0   
1888                 0.0                 0.0                 0.0   

      DISEASE GROUPING 4  DISEASE GROUPING 5  DISEASE GROUPING 6  ...  \
0                    0.0                 1.0                 1.0  ...   
9                    0.0                 0.0                 0.0  ...   
14                   0.0                 0.0                 0.0  ...   
19                   0.0                 0.0                 0.0  ...   
24                   0.0                 0.0                 0.0  ...   
...                  ...                 ...                 ...  ...   
1868                 0.0                 0.0                 0.0  ...   
1873                 0.0                 0.0                 0.0  ...   
1878                 0.0                 0.0                 0.0  ...   
1883                 0.0                 0.0                 0.0  ...   
1888                 0.0                 0.0                 0.0  ...   

      OXYGEN_SATURATION_DIFF  BLOODPRESSURE_DIASTOLIC_DIFF_REL  \
0                  -1.000000                         -1.000000   
9                  -0.959596                         -0.515528   
14                 -1.000000                         -1.000000   
19                 -0.979798                         -1.000000   
24                 -0.979798                         -0.860870   
...                      ...                               ...   
1868               -1.000000                         -1.000000   
1873               -1.000000                         -1.000000   
1878               -1.000000                         -1.000000   
1883               -1.000000                         -1.000000   
1888               -1.000000                         -1.000000   

      BLOODPRESSURE_SISTOLIC_DIFF_REL  HEART_RATE_DIFF_REL  \
0                           -1.000000            -1.000000   
9                           -0.351328            -0.747001   
14                          -1.000000            -1.000000   
19                          -0.883669            -0.956805   
24                          -0.714460            -0.986481   
...                               ...                  ...   
1868                        -1.000000            -1.000000   
1873                        -1.000000            -1.000000   
1878                        -1.000000            -1.000000   
1883                        -1.000000            -1.000000

In [13]:
features = ['AGE_PERCENTIL', 'GENDER'] + comorb_lst + lab_lst + vitalSigns_lst
X = training_data[features]
y = training_data['ICU_NEW']

In [14]:
SEED = 440
experiment = setup(
    training_data, 
    target='ICU_NEW',
    ignore_features=['PATIENT_VISIT_IDENTIFIER', 'ICU', 'WINDOW'],
    #+lab_columns_to_ignore,
    #fix_imbalance=True, # fixing train-test split imbalances
    #feature_selection=True, feature_selection_threshold=0.95, # conservative important feature selection
    #remove_perfect_collinearity=True, # in case we missed any perfectly collinear features
    session_id=SEED, # seed for reproductibility
    #silent=True # for kaggle compatibility
    )


 
Setup Succesfully Completed!


In [15]:
compare_models()

PyCaret's recommended experiment workflow is to use compare_models() right after setup to evaluate top performing models and finalize a few candidates for continued experimentation. As such, the function that actually allows to you create a model is unimaginatively called create_model(). This function creates a model and scores it using stratified cross-validation. Similar to compare_models(), the output prints a score grid that shows Accuracy, Recall, Precision, F1 and Kappa by fold.

For the remaining part of this project, we will work with the top 5 models as our candidate models. The selections are for illustration purposes only and do not necessarily mean they are the top performing or ideal for this type of data.

- Extra Trees Classifier('et')
- CatBoost Classifier('catboost')
- Random Forest Classifier('rf')
- Logistic Regression('lr')
- Extreme Gradient Boosting('xgboost')

##### Extra Trees Classifier

In [35]:
et = create_model('et')

Accuracy     AUC  Recall   Prec.      F1   Kappa
0       0.6800  0.7276  0.5000  0.7500  0.6000  0.3506
1       0.8000  0.9423  0.8333  0.7692  0.8000  0.6006
2       0.7200  0.8397  0.5833  0.7778  0.6667  0.4337
3       0.7600  0.8077  0.8333  0.7143  0.7692  0.5223
4       0.7200  0.7435  0.5455  0.7500  0.6316  0.4147
5       0.8000  0.8117  0.7273  0.8000  0.7619  0.5902
6       0.7200  0.8312  0.7273  0.6667  0.6957  0.4373
7       0.6250  0.6399  0.6364  0.5833  0.6087  0.2500
8       0.6250  0.7517  0.4545  0.6250  0.5263  0.2286
9       0.5417  0.6154  0.5455  0.5000  0.5217  0.0833
Mean    0.6992  0.7711  0.6386  0.6936  0.6582  0.3911
SD      0.0784  0.0923  0.1285  0.0927  0.0934  0.1575

In [36]:
tuned_et = tune_model('et')

Accuracy     AUC  Recall   Prec.      F1   Kappa
0       0.7200  0.6827  0.5833  0.7778  0.6667  0.4337
1       0.8800  0.9487  0.9167  0.8462  0.8800  0.7604
2       0.7200  0.8333  0.5833  0.7778  0.6667  0.4337
3       0.7600  0.8141  0.7500  0.7500  0.7500  0.5192
4       0.8000  0.7597  0.7273  0.8000  0.7619  0.5902
5       0.7600  0.7922  0.7273  0.7273  0.7273  0.5130
6       0.7200  0.8052  0.6364  0.7000  0.6667  0.4262
7       0.5833  0.6224  0.5455  0.5455  0.5455  0.1608
8       0.6250  0.7902  0.4545  0.6250  0.5263  0.2286
9       0.5000  0.6224  0.5455  0.4615  0.5000  0.0069
Mean    0.7068  0.7671  0.6470  0.7011  0.6691  0.4073
SD      0.1047  0.0955  0.1273  0.1154  0.1128  0.2089

In [41]:
predict_model(tuned_et);

Model  Accuracy     AUC  Recall   Prec.      F1   Kappa
0  Extra Trees Classifier    0.7453  0.7476  0.7551  0.7115  0.7327  0.4898

##### CatBoost Classifier

In [43]:
catboost = create_model('catboost')

Accuracy     AUC  Recall   Prec.      F1   Kappa
0       0.6800  0.8077  0.5000  0.7500  0.6000  0.3506
1       0.8400  0.8654  0.9167  0.7857  0.8462  0.6815
2       0.7200  0.7949  0.6667  0.7273  0.6957  0.4373
3       0.8000  0.8013  0.8333  0.7692  0.8000  0.6006
4       0.7600  0.8377  0.6364  0.7778  0.7000  0.5033
5       0.7600  0.8442  0.6364  0.7778  0.7000  0.5033
6       0.8000  0.8312  0.7273  0.8000  0.7619  0.5902
7       0.5833  0.6713  0.5455  0.5455  0.5455  0.1608
8       0.7083  0.8112  0.6364  0.7000  0.6667  0.4085
9       0.5833  0.6503  0.5455  0.5455  0.5455  0.1608
Mean    0.7235  0.7915  0.6644  0.7179  0.6861  0.4397
SD      0.0833  0.0686  0.1240  0.0906  0.0960  0.1672

In [44]:
tuned_catboost = tune_model('catboost')

Accuracy     AUC  Recall   Prec.      F1   Kappa
0       0.7200  0.8205  0.5000  0.8571  0.6316  0.4300
1       0.8800  0.9038  1.0000  0.8000  0.8889  0.7619
2       0.7600  0.7949  0.6667  0.8000  0.7273  0.5161
3       0.8400  0.8397  0.8333  0.8333  0.8333  0.6795
4       0.7200  0.8247  0.5455  0.7500  0.6316  0.4147
5       0.7600  0.8247  0.6364  0.7778  0.7000  0.5033
6       0.6800  0.8117  0.6364  0.6364  0.6364  0.3506
7       0.5833  0.6503  0.5455  0.5455  0.5455  0.1608
8       0.7083  0.8462  0.6364  0.7000  0.6667  0.4085
9       0.5417  0.6084  0.5455  0.5000  0.5217  0.0833
Mean    0.7193  0.7925  0.6545  0.7200  0.6783  0.4309
SD      0.0977  0.0865  0.1456  0.1162  0.1094  0.1962

In [45]:
predict_model(tuned_catboost);

Model  Accuracy     AUC  Recall   Prec.      F1   Kappa
0  CatBoost Classifier    0.6981  0.7479  0.6735  0.6735  0.6735  0.3928

##### Random Forest Classifier

In [46]:
rf = create_model('rf')

Accuracy     AUC  Recall   Prec.      F1   Kappa
0       0.6000  0.6571  0.3333  0.6667  0.4444  0.1830
1       0.7200  0.7660  0.8333  0.6667  0.7407  0.4444
2       0.6800  0.7404  0.5000  0.7500  0.6000  0.3506
3       0.8000  0.7788  0.7500  0.8182  0.7826  0.5981
4       0.7200  0.7630  0.5455  0.7500  0.6316  0.4147
5       0.7200  0.7955  0.6364  0.7000  0.6667  0.4262
6       0.6400  0.7305  0.5455  0.6000  0.5714  0.2623
7       0.6250  0.6783  0.5455  0.6000  0.5714  0.2394
8       0.5833  0.6993  0.3636  0.5714  0.4444  0.1367
9       0.5833  0.6119  0.5455  0.5455  0.5455  0.1608
Mean    0.6672  0.7221  0.5598  0.6668  0.5999  0.3216
SD      0.0688  0.0561  0.1453  0.0840  0.1056  0.1420

In [47]:
tuned_rf = tune_model('rf')

Accuracy     AUC  Recall   Prec.      F1   Kappa
0       0.7600  0.7885  0.5833  0.8750  0.7000  0.5130
1       0.8000  0.9038  0.8333  0.7692  0.8000  0.6006
2       0.6800  0.8077  0.7500  0.6429  0.6923  0.3631
3       0.7200  0.8269  0.7500  0.6923  0.7200  0.4409
4       0.7200  0.7143  0.5455  0.7500  0.6316  0.4147
5       0.6400  0.7468  0.4545  0.6250  0.5263  0.2475
6       0.7600  0.8896  0.7273  0.7273  0.7273  0.5130
7       0.5833  0.6364  0.5455  0.5455  0.5455  0.1608
8       0.6250  0.7972  0.4545  0.6250  0.5263  0.2286
9       0.5833  0.6294  0.6364  0.5385  0.5833  0.1724
Mean    0.6872  0.7741  0.6280  0.6791  0.6453  0.3655
SD      0.0729  0.0889  0.1256  0.0994  0.0915  0.1479

In [48]:
predict_model(tuned_rf);

Model  Accuracy     AUC  Recall   Prec.      F1  Kappa
0  Random Forest Classifier     0.717  0.7476  0.7347  0.6792  0.7059  0.434

##### Logistic Regression

In [49]:
lr = create_model('lr')

Accuracy     AUC  Recall   Prec.      F1   Kappa
0       0.6800  0.7821  0.6667  0.6667  0.6667  0.3590
1       0.8000  0.8205  0.7500  0.8182  0.7826  0.5981
2       0.7200  0.8462  0.5833  0.7778  0.6667  0.4337
3       0.6000  0.7244  0.4167  0.6250  0.5000  0.1883
4       0.7600  0.7987  0.6364  0.7778  0.7000  0.5033
5       0.6800  0.7338  0.5455  0.6667  0.6000  0.3377
6       0.6400  0.7662  0.5455  0.6000  0.5714  0.2623
7       0.6667  0.7063  0.7273  0.6154  0.6667  0.3379
8       0.6667  0.8042  0.6364  0.6364  0.6364  0.3287
9       0.6250  0.6923  0.4545  0.6250  0.5263  0.2286
Mean    0.6838  0.7675  0.5962  0.6809  0.6317  0.3578
SD      0.0580  0.0489  0.1033  0.0756  0.0801  0.1187

In [50]:
tuned_lr = tune_model('lr')

Accuracy     AUC  Recall   Prec.      F1   Kappa
0       0.6800  0.7756  0.6667  0.6667  0.6667  0.3590
1       0.7600  0.8205  0.7500  0.7500  0.7500  0.5192
2       0.7200  0.8526  0.5833  0.7778  0.6667  0.4337
3       0.6000  0.7308  0.5000  0.6000  0.5455  0.1935
4       0.8000  0.7987  0.7273  0.8000  0.7619  0.5902
5       0.7200  0.7403  0.5455  0.7500  0.6316  0.4147
6       0.6800  0.7727  0.5455  0.6667  0.6000  0.3377
7       0.6667  0.6853  0.8182  0.6000  0.6923  0.3469
8       0.6667  0.8042  0.6364  0.6364  0.6364  0.3287
9       0.6250  0.6993  0.4545  0.6250  0.5263  0.2286
Mean    0.6918  0.7680  0.6227  0.6872  0.6477  0.3752
SD      0.0569  0.0509  0.1115  0.0716  0.0735  0.1145

In [51]:
predict_model(tuned_lr);

Model  Accuracy     AUC  Recall   Prec.      F1   Kappa
0  Logistic Regression    0.7075  0.7698  0.6327  0.7045  0.6667  0.4075

##### Extreme Gradient Boosting

In [52]:
xgboost = create_model('xgboost')

Accuracy     AUC  Recall   Prec.      F1   Kappa
0       0.6400  0.7500  0.4167  0.7143  0.5263  0.2671
1       0.7600  0.8590  0.8333  0.7143  0.7692  0.5223
2       0.7200  0.8077  0.6667  0.7273  0.6957  0.4373
3       0.8000  0.7885  0.8333  0.7692  0.8000  0.6006
4       0.7600  0.7597  0.6364  0.7778  0.7000  0.5033
5       0.7200  0.8052  0.5455  0.7500  0.6316  0.4147
6       0.7200  0.8247  0.7273  0.6667  0.6957  0.4373
7       0.5833  0.6643  0.5455  0.5455  0.5455  0.1608
8       0.7500  0.7552  0.7273  0.7273  0.7273  0.4965
9       0.6667  0.7063  0.5455  0.6667  0.6000  0.3191
Mean    0.7120  0.7721  0.6477  0.7059  0.6691  0.4159
SD      0.0615  0.0546  0.1292  0.0640  0.0865  0.1253

In [53]:
tuned_xgboost = tune_model('xgboost')

Accuracy     AUC  Recall   Prec.      F1   Kappa
0       0.6800  0.7372  0.5000  0.7500  0.6000  0.3506
1       0.8400  0.8718  0.9167  0.7857  0.8462  0.6815
2       0.7600  0.8462  0.7500  0.7500  0.7500  0.5192
3       0.7200  0.7628  0.7500  0.6923  0.7200  0.4409
4       0.7600  0.7727  0.7273  0.7273  0.7273  0.5130
5       0.6000  0.7662  0.3636  0.5714  0.4444  0.1554
6       0.8000  0.7922  0.8182  0.7500  0.7826  0.5981
7       0.6250  0.6643  0.5455  0.6000  0.5714  0.2394
8       0.7500  0.7692  0.7273  0.7273  0.7273  0.4965
9       0.5833  0.6084  0.5455  0.5455  0.5455  0.1608
Mean    0.7118  0.7591  0.6644  0.6899  0.6715  0.4156
SD      0.0824  0.0734  0.1599  0.0811  0.1184  0.1730

In [54]:
predict_model(tuned_xgboost);

Model  Accuracy     AUC  Recall   Prec.     F1   Kappa
0  Extreme Gradient Boosting    0.7075  0.7351  0.7347  0.6667  0.699  0.4159

## PyCaret with the dataframe Standarized and Normalized

In [26]:
from sklearn.preprocessing import StandardScaler, Normalizer
from sklearn.pipeline import make_pipeline

pipeline = [
    StandardScaler(),
    Normalizer()
]

tr = make_pipeline(*pipeline)

In [27]:
data_prueba = raw_data_joined.iloc[:,4:229]

X_ready = tr.fit_transform(data_prueba)
X_ready = pd.DataFrame(X_ready, columns=data_prueba.columns)

data_prueba_dem = raw_data_joined.iloc[:,0:4]

data_prueba_pycaret = data_prueba_dem.join(X_ready)

aux_df_pycaret = raw_data_joined.iloc[:,229:231]

data_prueba_pycaret = data_prueba_pycaret.join(aux_df_pycaret)

In [28]:
data_prueba_pycaret

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  AGE_PERCENTIL  GENDER  \
0                            0            1             60       0   
1                            0            1             60       0   
2                            0            1             60       0   
3                            0            1             60       0   
4                            0            1             60       0   
...                        ...          ...            ...     ...   
1920                       384            0             50       1   
1921                       384            0             50       1   
1922                       384            0             50       1   
1923                       384            0             50       1   
1924                       384            0             50       1   

      DISEASE GROUPING 1  DISEASE GROUPING 2  DISEASE GROUPING 3  \
0              -0.041853           -0.020429           -0.039560   
1              -0.037514           -0.018311           -0.035459   
2              -0.037514           -0.018311           -0.035459   
3              -0.041088           -0.020055           -0.038837   
4              -0.018762           -0.009158           -0.017734   
...                  ...                 ...                 ...   
1920           -0.036946           -0.018033           -0.034922   
1921           -0.038498           -0.018791           -0.036389   
1922           -0.039982           -0.019516           -0.037792   
1923           -0.035692           -0.017422           -0.033737   
1924           -0.049557           -0.024189            0.432790   

      DISEASE GROUPING 4  DISEASE GROUPING 5  DISEASE GROUPING 6  ...  \
0              -0.017064            0.314148            0.542968  ...   
1              -0.015295            0.281580            0.486677  ...   
2              -0.015295            0.281580            0.486677  ...   
3              -0.016752            0.308409            0.533048  ...   
4              -0.007649            0.140825            0.243399  ...   
...                  ...                 ...                 ...  ...   
1920           -0.015063           -0.040631           -0.023508  ...   
1921           -0.015696           -0.042339           -0.024496  ...   
1922           -0.016302           -0.043971           -0.025440  ...   
1923           -0.014552           -0.039253           -0.022711  ...   
1924           -0.020205           -0.054500           -0.031533  ...   

      TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  \
0            -0.070432               -0.039602   
1            -0.063130               -0.035497   
2            -0.063130               -0.035497   
3            -0.069145               -0.038879   
4             0.111539                0.022440   
...                ...                     ...   
1920         -0.062174               -0.034959   
1921         -0.064787               -0.036428   
1922         -0.067284               -0.037833   
1923         -0.060065               -0.033773   
1924          0.141050                0.047477   

      BLOODPRESSURE_DIASTOLIC_DIFF_REL  BLOODPRESSURE_SISTOLIC_DIFF_REL  \
0                            -0.066468                        -0.067127   
1                            -0.059577                        -0.060168   
2                            -0.059577                        -0.060168   
3                            -0.065254                        -0.065901   
4                             0.086509                         0.178283   
...                                ...                              ...   
1920                         -0.058675                        -0.059257   
1921                         -0.061141                        -0.061747   
1922                         -0.063498                        -0.064127   
1923                         -0.056685                        -0.057247   
1924                          0.071435       

In [29]:
train_data_normalized = data_prueba_pycaret.loc[~((data_prueba_pycaret['WINDOW'] == '0-2') & (data_prueba_pycaret['ICU'] == 1))]

icu_above_2_normalized = train_data_normalized.groupby('PATIENT_VISIT_IDENTIFIER')\
    .agg({'ICU': max})\
    .reset_index()\
    .rename(columns={'ICU': 'ICU_NEW'})

training_data_normalized = train_data_normalized.merge(icu_above_2_normalized, on=['PATIENT_VISIT_IDENTIFIER'], how='left')

mask_02 = training_data_normalized.WINDOW == '0-2'

training_data_normalized = training_data_normalized.loc[mask_02]

In [30]:
training_data_normalized

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  AGE_PERCENTIL  GENDER  \
0                            0            1             60       0   
9                            2            0             10       0   
14                           3            0             40       1   
19                           4            0             10       0   
24                           5            0             10       0   
...                        ...          ...            ...     ...   
1868                       380            0             40       1   
1873                       381            1             90       0   
1878                       382            0             50       0   
1883                       383            0             40       1   
1888                       384            0             50       1   

      DISEASE GROUPING 1  DISEASE GROUPING 2  DISEASE GROUPING 3  \
0              -0.041853           -0.020429           -0.039560   
9              -0.030023           -0.014654           -0.028378   
14             -0.021905           -0.010692           -0.020705   
19             -0.039387           -0.019225           -0.037230   
24             -0.038380           -0.018733           -0.036277   
...                  ...                 ...                 ...   
1868           -0.009744           -0.004756           -0.009211   
1873           -0.047333           -0.023104           -0.044740   
1878           -0.040780           -0.019905           -0.038546   
1883           -0.059346           -0.028967           -0.056095   
1888           -0.036946           -0.018033           -0.034922   

      DISEASE GROUPING 4  DISEASE GROUPING 5  DISEASE GROUPING 6  ...  \
0              -0.017064            0.314148            0.542968  ...   
9              -0.012241           -0.033018           -0.019103  ...   
14             -0.008931           -0.024090           -0.013938  ...   
19             -0.016059           -0.043316           -0.025062  ...   
24             -0.015648           -0.042209           -0.024421  ...   
...                  ...                 ...                 ...  ...   
1868           -0.003973           -0.010716           -0.006200  ...   
1873           -0.019299           -0.052055           -0.030118  ...   
1878           -0.016627           -0.044848           -0.025948  ...   
1883           -0.024197           -0.065266           -0.037762  ...   
1888           -0.015063           -0.040631           -0.023508  ...   

      OXYGEN_SATURATION_DIFF  BLOODPRESSURE_DIASTOLIC_DIFF_REL  \
0                  -0.039602                         -0.066468   
9                  -0.014116                          0.100127   
14                 -0.020727                         -0.034788   
19                 -0.027894                         -0.062553   
24                 -0.027181                         -0.006690   
...                      ...                               ...   
1868               -0.009220                         -0.015476   
1873               -0.044788                         -0.075172   
1878               -0.038588                         -0.064765   
1883               -0.056155                         -0.094251   
1888               -0.034959                         -0.058675   

      BLOODPRESSURE_SISTOLIC_DIFF_REL  HEART_RATE_DIFF_REL  \
0                           -0.067127            -0.066671   
9                            0.105517             0.044716   
14                          -0.035133            -0.034895   
19                          -0.027018            -0.042016   
24                           0.024916            -0.054818   
...                               ...                  ...   
1868                        -0.015629            -0.015523   
1873                        -0.075918            -0.075402   
1878                        -0.065407            -0.064963   
1883                        -0.095185            -0.094538

In [31]:
features = ['AGE_PERCENTIL', 'GENDER'] + comorb_lst + lab_lst + vitalSigns_lst
X_normalized = training_data_normalized[features]
y_normalized = training_data_normalized['ICU_NEW']

X_train_n, X_test_n, y_train_n, y_test_n = train_test_split(X_normalized, y_normalized, test_size=0.2, 
                                                  random_state=0)

In [32]:
SEED = 440
experiment = setup(
    training_data_normalized, 
    target='ICU_NEW',
    ignore_features=['PATIENT_VISIT_IDENTIFIER', 'ICU', 'WINDOW'],
    #+lab_columns_to_ignore,
    #fix_imbalance=True, # fixing train-test split imbalances
    #feature_selection=True, feature_selection_threshold=0.95, # conservative important feature selection
    #remove_perfect_collinearity=True, # in case we missed any perfectly collinear features
    session_id=SEED, # seed for reproductibility
    #silent=True # for kaggle compatibility
    )

 
Setup Succesfully Completed!


In [33]:
compare_models()

For the remaining part of this project, we will work with the top 5 models as our candidate models. The selections are for illustration purposes only and do not necessarily mean they are the top performing or ideal for this type of data.

- Extra Trees Classifier('et')
- CatBoost Classifier('catboost')
- Random Forest Classifier('rf')
- Logistic Regression('lr')
- Extreme Gradient Boosting('xgboost')